### PR TITLE
norman: streamline-frame wall-shear targets to attack τ_y/τ_z gap

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Rotating the wall-shear training targets into a local streamline-aligned frame will reduce the τ_y/τ_z gap (currently 2.35× and 2.73× above AB-UPT target) by giving the model a representation where the physically significant cross-flow components are explicitly separated from the dominant streamwise component.**

In the current Cartesian frame, τ_x dominates (~90% of wall-shear magnitude along the body). τ_y and τ_z are near-zero everywhere except at separation lines, vortex footprints, and wake-adjacent regions. MSE-trained models in this regime learn to predict τ_y ≈ 0 and τ_z ≈ 0 almost everywhere, because that is the globally optimal MSE answer. The result: systematic under-prediction of the physically important cross-flow events.

**Proposed fix:** per-surface-point rotation of the 3-component wall-shear target `(τ_x, τ_y, τ_z)` into a local frame aligned with the freestream projection onto the surface tangent plane:

- `t1 = normalize(e_x - (e_x·n)n)` — streamwise surface tangent (projection of freestream `e_x=(1,0,0)` onto tangent plane)
- `t2 = n × t1` — cross-flow surface tangent (perpendicular to t1 in tangent plane)
- The rotated targets are `(τ_t1, τ_t2, τ_n)` where **τ_n ≈ 0** by physics (wall shear is tangential by definition)

In this frame:
- `τ_t1` captures the dominant streamwise shear (smooth, high SNR, easy to learn)
- `τ_t2` captures the cross-flow shear (physically important, more distinctive signal/noise ratio than Cartesian τ_y/τ_z)
- `τ_n` is a near-zero regularization target — its loss serves as a sanity check

The model predicts in this rotated frame during training; predictions are rotated BACK to Cartesian for validation and test metrics (so the AB-UPT metric is always computed in Cartesian space). At inference, apply the same inverse rotation.

This is different from PR #312/419 which used surface tangent vectors as extra INPUT features. This PR rotates the OUTPUT TARGET space.

### Step 1 — Add CLI flag

```python
parser.add_argument('--wallshear-frame', type=str, default='cartesian',
    choices=['cartesian', 'streamline'],
    help='Target frame for wall-shear prediction. "streamline" rotates (tau_x, tau_y, tau_z) targets into local (t1, t2, n) frame per surface point.')
```

### Step 2 — Precompute the per-point rotation frame

In the training/validation loop, after loading the batch surface data (shape `[B, N_surf, C]`), compute the local frame. Surface normals are in `surface_x[:, :, 3:6]` (channels 3, 4, 5):

```python
def compute_streamline_frame(normals):
    """
    normals: [B, N_surf, 3]
    Returns t1, t2: [B, N_surf, 3] — streamwise and crossflow tangents
    """
    e_x = torch.zeros_like(normals)
    e_x[..., 0] = 1.0  # freestream direction (1, 0, 0)
    # t1: freestream projected onto tangent plane
    dot = (e_x * normals).sum(dim=-1, keepdim=True)  # [B, N, 1]
    t1 = e_x - dot * normals
    t1 = t1 / (t1.norm(dim=-1, keepdim=True) + 1e-8)
    # t2: cross-flow tangent
    t2 = torch.linalg.cross(normals, t1, dim=-1)
    t2 = t2 / (t2.norm(dim=-1, keepdim=True) + 1e-8)
    return t1, t2

def rotate_to_streamline(ws, t1, t2, normals):
    """
    ws: [B, N_surf, 3] — Cartesian (tau_x, tau_y, tau_z)
    Returns: [B, N_surf, 3] — (tau_t1, tau_t2, tau_n)
    """
    tau_t1 = (ws * t1).sum(dim=-1, keepdim=True)
    tau_t2 = (ws * t2).sum(dim=-1, keepdim=True)
    tau_n  = (ws * normals).sum(dim=-1, keepdim=True)
    return torch.cat([tau_t1, tau_t2, tau_n], dim=-1)

def rotate_from_streamline(ws_rot, t1, t2, normals):
    """
    Inverse: ws_rot [B, N_surf, 3] -> Cartesian [B, N_surf, 3]
    """
    return (ws_rot[..., 0:1] * t1 +
            ws_rot[..., 1:2] * t2 +
            ws_rot[..., 2:3] * normals)
```

### Step 3 — Apply rotation to targets in training loop

Wall-shear targets are in `surface_y[:, :, 1:4]` (channels 1, 2, 3 = τ_x, τ_y, τ_z). When `--wallshear-frame streamline`:

```python
if args.wallshear_frame == 'streamline':
    normals = surface_x[:, :, 3:6]  # [B, N_surf, 3]
    t1, t2 = compute_streamline_frame(normals)
    ws_cart = surface_y[:, :, 1:4]  # [B, N_surf, 3]
    ws_rot = rotate_to_streamline(ws_cart, t1, t2, normals)
    surface_y_train = surface_y.clone()
    surface_y_train[:, :, 1:4] = ws_rot
else:
    surface_y_train = surface_y
```

Apply the same rotation to model predictions BEFORE loss computation (so loss is in the rotated frame). Log `train/loss_ws_n` (the τ_n component loss) separately — it should start high and decay to near-zero by epoch 2, confirming the rotation is physically correct.

### Step 4 — Inverse rotation at validation

At validation/test time, rotate predictions BACK to Cartesian before computing AB-UPT metrics:

```python
if args.wallshear_frame == 'streamline':
    pred_ws_rot = pred_surface[:, :, 1:4]
    pred_ws_cart = rotate_from_streamline(pred_ws_rot, t1, t2, normals)
    pred_surface_eval = pred_surface.clone()
    pred_surface_eval[:, :, 1:4] = pred_ws_cart
else:
    pred_surface_eval = pred_surface
```

Then compute val metrics on `pred_surface_eval` as usual.

### Step 5 — Extra diagnostics

Log per-step:
- `train/loss_ws_t1` — streamwise shear loss (dominant component)
- `train/loss_ws_t2` — cross-flow shear loss (physically important)
- `train/loss_ws_n` — normal component loss (should → 0 by epoch 2)
- At epoch end: `val/ws_t2_rel_l2_pct` in the rotated frame (to confirm learning signal on cross-flow)

### Step 6 — Sweep (2 arms, sequential on 4×H100 DDP)

**Arm A (control):** `--wallshear-frame cartesian` (standard Cartesian targets)
**Arm B:** `--wallshear-frame streamline` (rotated frame + inverse at val)

Use `--wandb-group norman-r34-streamline-frame` for both arms.

**Full launch command for Arm A:**
```bash
cd target/
nohup torchrun --standalone --nproc_per_node=4 train.py \
  --agent norman \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --clip-grad-norm 0.5 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --learnable-pe \
  --no-compile-model \
  --wallshear-frame cartesian \
  --epochs 3 \
  --wandb-group norman-r34-streamline-frame \
  --wandb-name norman-r34-arm-a-cartesian \
  > _runs/arm_a_cartesian.log 2>&1 &
echo "PID: $!"
```

Replace `--wallshear-frame cartesian` with `streamline` for Arm B.

### IMPORTANT flag check

Use `--clip-grad-norm 0.5` (NOT `--grad-clip-norm`). Verify:
```bash
python train.py --help | grep -i clip
```

Also verify the normal/tangent channel indices are correct for this version of the code:
```bash
python -c "
import torch, sys
sys.path.insert(0, '.')
# quick sanity: load one case, print surface_x shape and channel meanings
"
```

Or just log `surface_x[:1, :5, 3:6].mean()` at step 0 to confirm normals are non-trivial.

## Results to report

For each arm, post:
1. W&B run id + link
2. Per-epoch `val/abupt_axis_mean_rel_l2_pct` table
3. Per-channel at best epoch: `surface_pressure`, `tau_y`, `tau_z`, `volume_pressure`
4. Arm B only: `train/loss_ws_n` at epochs 1, 2, 3 (should trend toward 0)
5. `test/abupt_axis_mean_rel_l2_pct` from best-val checkpoint

Win criterion: **val_abupt < 9.032%** AND Arm B τ_y or τ_z shows ≥10% relative improvement vs Arm A. The physical signal check (τ_n → 0) is a required diagnostic regardless of main metric outcome — report it explicitly.

## Baseline

Current best on yi:
- **val_abupt: 9.032%** (PR #517 askeladd, W&B run `brat65z4`, Lion lr=1e-4 clip=0.5)
- STRING-sep learnable PE now on yi via PR #490; frieren resumed run reached **8.087%** with `--learnable-pe` (awaiting clean from-scratch confirmation from PR #539)
- Expected next bar: ~8.1–8.5% once PR #539 lands

AB-UPT reference targets (what we're trying to match):
- τ_y (ws_y): 3.65% — **current yi: ~8.58%, gap 2.35×**
- τ_z (ws_z): 3.63% — **current yi: ~9.93%, gap 2.73×**
- surface_pressure: 3.82%
- volume_pressure: 6.08%
- wall_shear (overall): 7.29%

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #564 because ADVISOR closing this PR immediately — duplicate hypothesis. Askeladd is already running the streamline-aligned wall-shear target frame experiment in PR #541 (currently mid-sweep with Arm B running). I made this assignment without fully checking PR #541's status first — apologies for the confusion. A new distinct experiment will be assigned to you shortly..
